### ISSUE TITLE
Bug in Env.stdin_install

### ISSUE THREAD

**Issue by sydow** [actonlang/acton#1719]
### Acton Version

0.19.2.20240305.11.55.57

### Steps to Reproduce and Observed Behavior

Compile and run the following program
```
actor main(env):
    def interact(input : str):
        print(input)
    env.stdin_install(interact)
```
The ouput is 
```
examples> ./harmonic
123
b'123\n'
```

The input, which is supposed to be str, is a bytes object. The problem seems to be in base/builtin/env.c, where stdin_installG_local calls  uv_read_start with read_stdin as a parameter. read_stdin, in turn, hands the read chars to to$bytesD_len, which produces a bytes object.

### Expected Behavior

examples> ./harmonic
123
'123\n'

**Comment by plajjan** [comment]
I think I changed this deliberately. Our str are unicode/utf-8 and I don't think we can blindly assume stdin/stdout to be utf-8.

**Comment by sydow** [comment]
And I don’t think we don’t want __builtin__.act to say that the parameter to interact is a str and the runtime to give us a bytes object.

I also don’t want a bytes object when interacting with stdin. I much prefer to assume utf-8 and get a str.

Björn

**Comment by plajjan** [comment]
@sydow the type signature should obviously align :) so that's a bug and I presume a mistake on my part.

I would also prefer utf-8 when interacting with stdin but it's technically wrong. Not all terminals are utf-8 so I'm not sure what to do about this really. Using bytes means the Acton implementation is 100% correct although less convenient. Assuming utf-8 means it will break on anything that is not unicode. We could have heuristics but we don't support other encodings, so this certainly is not straightforward. I do prefer that the users application make the assumption about unicode terminal rather than the Acton Env. Just tack on a .decode() in your program and you're good to go? :)

**Comment by sydow** [comment]
So I tack on a decode() which can only decode to utf-8? Well, I don’t like it..

**Comment by plajjan** [comment]
@sydow I understand you do not like it. I don't like it either. This is a case of shifting the problem from the programming language to the application programmer. It's bad but I still think it's better that Acton is correct than break in various ways for non-unicode terminals, no? :)

**Comment by sydow** [comment]
> On 5 Mar 2024, at 15:36, Kristian Larsson ***@***.***> wrote:
> 
> 
> @sydow I understand you do not like it. I don't like it either. This is a case of shifting the problem from the programming language to the application programmer. It's bad but I still think it's better that Acton is correct than break in various ways for non-unicode terminals, no? :) 

No.

How do you mean that acton (after fixing __builtin__.act to give bytes as the type of interact’s  parameter) would correctly handle non-utf8 terminals? By letting its users communicate in bytes? This, I think, would be the only  “correct” solution, but it is useless. We are not shifting the problem to the application programmers, since they cannot do anything at all to solve it. Or do you expect the application programmers to write an encoding function to their preferred encoding? We don’t provide any. We simply cannot handle non-utf8 terminals.

It is certainly a shortcoming that we don’t support any other encodings than utf-8, but it is a very small deficit compared to many others. I don’t expect that we will do anything about this in the short to medium term; do you?

I would accept having an optional second argument to interact where one can specify encoding (one couldn’t use it today since we can’t encode anything else than utf-8 which would be the default) or add another method set_encoding to Env (same comment), but I certainly want interact to have a str parameter. So far, I have only become more convinced :-)

Björn

**Comment by sydow** [comment]
Hmm, the method is called stdin_install, not interact...

**Comment by plajjan** [comment]
@sydow (agreed to fix __builtin__.act to give bytes type) and then Acton would not with it's built-in capabilities offer any way to work with non-utf8 terminals, but at least you could try to decode the bytes you received and get an error. Or (with some module support in acton) you could include a latin-1 encoding thingy from somewhere on the Internet. Or you could just give an error message saying "we don't support this non-utf8 codec".

We could change stding_install to str today and then change it back at some other point when we have more supported encodings, but that requires user programs to change now and later on. I think for things like this it's better to align on what we want the long term design to be.

How are we to send binary data over stdin if stdin_install assumes utf-8? I think we must support bytes. Is there some clever type trick to support both / either? A union of str|bytes? Is that easy to work with?

I had the same woes when I wrote the process module. Should stdout/stderr of the child process return decoded unicode strings? I decided against it in the end and went with bytes for the same reason. With Python's subprocess module you can tell it to decode using utf-8 or something else, which plays along nicely in Python's untyped world. In Acton, I don't know how that would look like, if it's a union don't we end up always matching on the union type? That feels more painful that just adding a decode layer. The decode/encode from strings to bytes in Python 3 is actually very explicit and I hated it at first but have come to understand why it is necessary.

It doesn't seem like a big thing to me to write some convenient wrapper around stdin_install that decodes the content and calls some other callback, no?

I'd prefer if we don't make convenience changes for the short or medium term that we know we want to change in the longer term. I mean, it's okay that we have shortcuts for short/medium term future because we don't have the time to implement some stuff... but in this case, I'd argue the proper solution is already in place with a way for the application developer to implement the program (use decode!), so unless I'm missing some more convenient way of supporting bytes input and the correctness that comes with specifying the encoding, I think we are where we should be :)

**Comment by sydow** [comment]
Kristian, I have neither time nor wish to quarrel with you about this issue, but
- I don’t think we have agreed to fix __builtin__.act
- I don’t think it is clear that "we are where we should be”.

We could talk about it at today’s meeting but hopefully not spend the whole meeting on this :-)

**Comment by plajjan** [comment]
@sydow the way I see it, the fact that __builtin__.act and __builtin__.ext.c do not agree on the type (str vs bytes) is a plain bug. The current intention of the code I wrote for this and what is in fact returned is bytes, so __builtin__.act should be changed to reflect that. I've submitted #1722 to fix that.

Separate to the bug fix, you want to change it to actually be str by doing an implicit unicode decode. For multiple reasons, I don't think we should do this. *I think* we are where we should be (it's subjective, not objective :) ). Yes, let's try to discuss this rather briefly in our meeting :)

**Comment by plajjan** [comment]
@sydow meeting soon.. but do you have a suggestion for a design to support using unicode that supports:
- reconfigurable encoding, so we can get raw bytes if we want to send say binary data
- error handling for unicode decoding errors

I reckon we could do `stdin_install(on_str_cb: ?action(str) -> None, encoding: ?str, on_bytes_cb: ?action(bytes) -> None)` and only one of `on_str_cb` or `on_bytes_cb` can be used at a time (or we throw exception). `encoding` argument dictates which encoding should be presumed, per default unicode and it could only support builtin encodings. Still not sure how we should handle a decoding error for `on_str_cb`. Maybe it should actually be `on_str_cb(data: ?str, error: ?str) -> None` ??

An alternative would be a single common function with keyword arguments for str vs bytes, like `on_stdin_cb(data: ?str, bdata: ?bytes, error: ?str)`.

Optional arguments are sort of bad because they require `if x is not None` checking but having separate arguments means we don't need `if isinstance(x, bytes)` or similar, which is arguably slightly better. The separate callbacks for str vs bytes reduce this problem but the error argument means we still need to check that `data` is a `str`. A separate error callback would reduce that... but is obviously another callback to manage. We don't have unions today but I think unions is bad because with a union you would end up always having to check what the returned type is. Or maybe that sort of checking is just normal and should be accepted?

**Comment by plajjan** [comment]
Fixed by #1723
